pkg/ccl/testccl/sqlccl/sqlccl_test: TestExplainGist failed [the first bucket should have NumRange=0]

offtopic


Jira issue: CRDB-47828

pkg/ccl/testccl/sqlccl/sqlccl_test.TestExplainGist [failed](https://tanzanite.cluster.engflow.com/invocations/default/e76f9225-be68-40a5-8c67-05dccd5e9817?testReportRun=6&testReportShard=2&testReportAttempt=1#targets-Ly9wa2cvY2NsL3Rlc3RjY2wvc3FsY2NsOnNxbGNjbF90ZXN0) on master @ [22b262749c502d07ec7ccec5b76abd67c361ae4d](https://github.com/cockroachdb/cockroach/commits/22b262749c502d07ec7ccec5b76abd67c361ae4d):


```
        		e' \'<p8p\x11k':::STRING,
        		'c39d0ff5-01ab-40c7-a41f-9b818e19ca5c':::UUID,
        		'175.86.186.138/5':::INET,
        		'[["baz"], null, 1.6975160350149279]':::JSONB
        	);
        INSERT
        INTO
        	defaultdb.public.seed AS tab_588
        		(_int4, _int8, _float4, _float8, _date, _timestamp, _interval, _bool, _bytes, _uuid, _inet, _jsonb, _enum)
        VALUES
        	(
        		(-257792830):::INT8,
        		NULL,
        		NULL,
        		2.2659187276958885:::FLOAT8,
        		'1979-07-29':::DATE,
        		'2006-01-03 14:50:08.000986':::TIMESTAMP,
        		'22 years 2 mons 114 days 01:41:18.595892':::INTERVAL,
        		true,
        		'\xb52bb344ae':::BYTES,
        		'6f88982c-f30e-4fb0-99dc-6a473ebe39e8':::UUID,
        		'150.199.87.77/19':::INET,
        		'[{"baz": "c", "foobar": true}, [false, 1.2722919345199613, false], [[{"baz": 0.37131715814737487}, true, false], 0.3426469686185628], null, true, null]':::JSONB,
        		'hello':::greeting
        	);
        UPDATE
        	defaultdb.public.seed AS tab_589
        SET
        	_jsonb = tab_589._jsonb, _int8 = tab_589._int8
        WHERE
        	false
        ORDER BY
        	tab_589._float4 DESC NULLS FIRST,
        	tab_589._timestamptz NULLS FIRST,
        	tab_589._uuid DESC,
        	tab_589.crdb_internal_mvcc_timestamp DESC NULLS LAST
        LIMIT
        	45:::INT8;
        
    explain_test.go:207: pq: internal error: the first bucket should have NumRange=0: UPDATE
        	defaultdb.public.seed AS tab_590
        SET
        	_decimal = (-2942154.205475587217):::DECIMAL
        FROM
        	defaultdb.public.seed AS tab_591, defaultdb.public.seed AS tab_592, defaultdb.public.seed AS tab_593
        WHERE
        	tab_592._bool
        LIMIT
        	40:::INT8
    --- FAIL: TestExplainGist/main (5.10s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=6</code>
 - <code>shard=2</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestExplainGist.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/roachpb/data.go
+++ b/pkg/roachpb/data.go
@@ -945,17 +945,17 @@ func (v Value) PrettyPrint() (ret string) {
 			if i != 0 {
 				buf.WriteRune('/')
 			}
-			_, _, colIDDiff, typ, err := encoding.DecodeValueTag(b)
+			_, _, colIDDelta, typ, err := encoding.DecodeValueTag(b)
 			if err != nil {
 				break
 			}
-			colID += colIDDiff
+			colID += colIDDelta
 			var s string
 			b, s, err = encoding.PrettyPrintValueEncoded(b)
 			if err != nil {
 				break
 			}
-			fmt.Fprintf(&buf, "%d:%d:%s/%s", colIDDiff, colID, typ, s)
+			fmt.Fprintf(&buf, "%d:%d:%s/%s", colIDDelta, colID, typ, s)
 		}
 	case ValueType_INT:
 		var i int64

--- a/pkg/sql/colfetcher/cfetcher.go
+++ b/pkg/sql/colfetcher/cfetcher.go
@@ -1271,7 +1271,7 @@ func (cf *cFetcher) processValueBytes(
 	cf.machine.remainingValueColsByIdx.UnionWith(cf.table.compositeIndexColOrdinals)
 
 	var (
-		colIDDiff      uint32
+		colIDDelta     uint32
 		lastColID      descpb.ColumnID
 		dataOffset     int
 		typ            encoding.Type
@@ -1283,11 +1283,11 @@ func (cf *cFetcher) processValueBytes(
 	// it's expensive to keep calling .Len() in the loop.
 	remainingValueCols := cf.machine.remainingValueColsByIdx.Len()
 	for len(valueBytes) > 0 && remainingValueCols > 0 {
-		_, dataOffset, colIDDiff, typ, err = encoding.DecodeValueTag(valueBytes)
+		_, dataOffset, colIDDelta, typ, err = encoding.DecodeValueTag(valueBytes)
 		if err != nil {
 			return "", "", err
 		}
-		colID := lastColID + descpb.ColumnID(colIDDiff)
+		colID := lastColID + descpb.ColumnID(colIDDelta)
 		lastColID = colID
 		vecIdx := -1
 		// Find the ordinal into table.cols for the column ID we just decoded,

--- a/pkg/sql/logictest/testdata/logic_test/stats
+++ b/pkg/sql/logictest/testdata/logic_test/stats
@@ -106,3 +106,19 @@ FROM [SHOW STATISTICS FOR TABLE t139381]
 {k}  histogram_collected
 {j}  histogram_collected
 {v}  histogram_collected
+
+# Regression test for inflating the distinct count due to including column ID
+# delta in the sketch (#141448).
+statement ok
+CREATE TABLE t141448 (i INT, f FLOAT, b BOOL);
+
+statement ok
+INSERT INTO t141448 VALUES (NULL, NULL, false), (NULL, 1, false), (1, 1, false);
+
+statement ok
+ANALYZE t141448;
+
+query II
+SELECT row_count, distinct_count FROM [SHOW STATISTICS FOR TABLE t141448] WHERE column_names = ARRAY['b'];
+----
+3  1

--- a/pkg/sql/rowenc/index_encoding.go
+++ b/pkg/sql/rowenc/index_encoding.go
@@ -494,11 +494,11 @@ func DecodeIndexKeyToDatums(
 
 		var lastColID descpb.ColumnID = 0
 		for len(valueBytes) > 0 {
-			typeOffset, dataOffset, colIDDiff, typ, err := encoding.DecodeValueTag(valueBytes)
+			typeOffset, dataOffset, colIDDelta, typ, err := encoding.DecodeValueTag(valueBytes)
 			if err != nil {
 				return nil, err
 			}
-			colID := lastColID + descpb.ColumnID(colIDDiff)
+			colID := lastColID + descpb.ColumnID(colIDDelta)
 			lastColID = colID
 			colOrdinal, ok := colIDs.Get(colID)
 			if !ok {
@@ -1824,16 +1824,16 @@ func SkipColumnNotInPrimaryIndexValue(
 func DecodeValueBytes(
 	colIdxMap catalog.TableColMap, valueBytes []byte, neededValueCols int, row EncDatumRow,
 ) (colOrds intsets.Fast, err error) {
-	var colIDDiff uint32
+	var colIDDelta uint32
 	var lastColID descpb.ColumnID
 	var typeOffset, dataOffset int
 	var typ encoding.Type
 	for len(valueBytes) > 0 && colOrds.Len() < neededValueCols {
-		typeOffset, dataOffset, colIDDiff, typ, err = encoding.DecodeValueTag(valueBytes)
+		typeOffset, dataOffset, colIDDelta, typ, err = encoding.DecodeValueTag(valueBytes)
 		if err != nil {
 			return intsets.Fast{}, err
 		}
-		colID := lastColID + descpb.ColumnID(colIDDiff)
+		colID := lastColID + descpb.ColumnID(colIDDelta)
 		lastColID = colID
 		idx, ok := colIdxMap.Get(colID)
 		if !ok {

--- a/pkg/sql/rowenc/valueside/decode.go
+++ b/pkg/sql/rowenc/valueside/decode.go
@@ -306,11 +306,11 @@ func (d *Decoder) Decode(a *tree.DatumAlloc, bytes []byte) (tree.Datums, error) 
 
 	var lastColID descpb.ColumnID
 	for len(bytes) > 0 {
-		_, dataOffset, colIDDiff, typ, err := encoding.DecodeValueTag(bytes)
+		_, dataOffset, colIDDelta, typ, err := encoding.DecodeValueTag(bytes)
 		if err != nil {
 			return nil, err
 		}
-		colID := lastColID + descpb.ColumnID(colIDDiff)
+		colID := lastColID + descpb.ColumnID(colIDDelta)
 		lastColID = colID
 		idx, ok := d.colIdxMap.Get(colID)
 		if !ok {

--- a/pkg/sql/rowexec/sampler.go
+++ b/pkg/sql/rowexec/sampler.go
@@ -14,6 +14,7 @@ import (
 	hllNew "github.com/axiomhq/hyperloglog"
 	hllOld "github.com/axiomhq/hyperloglog/000"
 	"github.com/cockroachdb/cockroach/pkg/server/telemetry"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/catenumpb"
 	"github.com/cockroachdb/cockroach/pkg/sql/execinfra"
 	"github.com/cockroachdb/cockroach/pkg/sql/execinfrapb"
 	"github.com/cockroachdb/cockroach/pkg/sql/execversion"
@@ -24,6 +25,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/stats"
 	"github.com/cockroachdb/cockroach/pkg/sql/types"
 	"github.com/cockroachdb/cockroach/pkg/util/collatedstring"
+	"github.com/cockroachdb/cockroach/pkg/util/encoding"
 	"github.com/cockroachdb/cockroach/pkg/util/intsets"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/mon"
@@ -520,18 +522,17 @@ func (s *sketchInfo) addRow(
 	}
 
 	s.numRows++
-	isNull := true
+	allNulls := true
 	*buf = (*buf)[:0]
 	for _, col := range s.spec.Columns {
-		if b := row[col].EncodedBytes(); b != nil && !containsCollatedString(typs[col]) {
-			// If the datum is already encoded and does not contain a collated
-			// string, we can insert the encoded bytes directly into the sketch.
-			// Even though the encoded bytes contain the column ID and type,
-			// this will not break the "loose invariant" that equal datums will
-			// have equal bytes added to the sketch because we are always adding
-			// datums from the same table column, i.e., the column ID and type
-			// should be the same for every value in the column.
-			//
+		// Avoid calling IsNull() if the datum is unset because it will panic.
+		// Instead, return an assertion error that might help in debugging.
+		if row[col].IsUnset() {
+			return errors.AssertionFailedf("unset datum: col=%d row=%s", col, row.String(typs))
+		}
+		isNull := row[col].IsNull()
+		allNulls = allNulls && isNull
+		if b := row[col].EncodedBytes(); b != nil && !containsCollatedString(typs[col]) && !isNull {
 			// Composite, value encoded datums may have different encodings for
 			// semantically equivalent types, so using the encoded bytes can
 			// skew the cardinality estimate slightly. This should be rare and
@@ -551,11 +552,33 @@ func (s *sketchInfo) addRow(
 			// "fOO", and "FOO". For this reason, we fall-back to using
 			// Fingerprint for collated strings.
 			//
+			// For NULL datums we use the Fingerprint method so that regardless
+			// of datum encoding, all of them got the same encoding.
+			//
 			// TODO(mgartner): We should probably truncate b to some max size to
 			// prevent a really wide value from growing buf. Since the distinct
 			// count is an estimate anyway, truncating the value shouldn't have
 			// any real impact.
-			*buf = append(*buf, b...)
+			if enc, _ := row[col].Encoding(); enc == catenumpb.DatumEncoding_VALUE {
+				// Value encoding includes column ID delta in the prefix which
+				// can differ based on the values of other columns within the
+				// same row (i.e. whether the previous columns had NULL or
+				// non-NULL values). To prevent this detail from artificially
+				// increasing the distinct estimate, we'll remove the column ID
+				// delta from encoding that we use for the current datum.
+				_, dataOffset, _, typ, err := encoding.DecodeValueTag(b)
+				if err != nil {
+					return err
+				}
+				// Including the value tag (i.e. the type) allows us to
+				// differentiate some non-NULL values (like integer 0) from NULL
+				// ones.
+				*buf = append(*buf, byte(typ))
+				*buf = append(*buf, b[dataOffset:]...)
+			} else {
+				// Key-encoded datums can be used as is.
+				*buf = append(*buf, b...)
+			}
 		} else {
 			// Fallback to using the Fingerprint method to generate bytes to
 			// insert into the sketch.
@@ -564,8 +587,8 @@ func (s *sketchInfo) addRow(
 			// independent (to prevent bounded memory leaks like we've seen in
 			// #136394). The problem in that issue was that the same backing
 			// slice of datums was shared across rows, so if a single row was
-			// kept as a sample, it could keep many garbage alive. To go around
-			// that we simply disabled the batching.
+			// kept as a sample, it could keep many garbage datums alive. To go
+			// around that we simply disabled the batching.
 			//
 			// We choose to not perform the memory accounting for possibly
 			// decoded tree.Datum because we will lose the references to row
@@ -575,19 +598,10 @@ func (s *sketchInfo) addRow(
 				return err
 			}
 		}
-		if isNull {
-			// Avoid calling IsNull() if the datum is unset because it will
-			// panic. Instead return an assertion error that might help in
-			// debugging.
-			if row[col].IsUnset() {
-				return errors.AssertionFailedf("unset datum: col=%d row=%s", col, row.String(typs))
-			}
-			isNull = row[col].IsNull()
-		}
 		s.size += int64(row[col].DiskSize())
 	}
 
-	if isNull {
+	if allNulls {
 		s.numNulls++
 	}
 	if s.sketchNew != nil {

--- a/pkg/util/encoding/encoding.go
+++ b/pkg/util/encoding/encoding.go
@@ -2553,8 +2553,8 @@ const NoColumnID uint32 = 0
 // methods.
 //
 // The prefix uses varints to encode a column id and type, packing them into a
-// single byte when they're small (colID < 8 and typ < 15). This works by
-// shifting the colID "left" by 4 and putting any type less than 15 in the low
+// single byte when they're small (colIDDelta < 8 and typ < 15). This works by
+// shifting the colIDDelta "left" by 4 and putting any type less than 15 in the low
 // bytes. The result is uvarint encoded and fits in one byte if the original
 // column id fit in 3 bits. If it doesn't fit in one byte, the most significant
 // bits spill to the "left", leaving the type bits always at the very "right".
@@ -2565,45 +2565,45 @@ const NoColumnID uint32 = 0
 //
 // Together, this means the everything but the last byte of the first uvarint
 // can be dropped if the column id isn't needed.
-func EncodeValueTag(appendTo []byte, colID uint32, typ Type) []byte {
+func EncodeValueTag(appendTo []byte, colIDDelta uint32, typ Type) []byte {
 	if typ >= SentinelType {
-		appendTo = EncodeNonsortingUvarint(appendTo, uint64(colID)<<4|uint64(SentinelType))
+		appendTo = EncodeNonsortingUvarint(appendTo, uint64(colIDDelta)<<4|uint64(SentinelType))
 		return EncodeNonsortingUvarint(appendTo, uint64(typ))
 	}
-	if colID == NoColumnID {
+	if colIDDelta == NoColumnID {
 		// TODO(dan): EncodeValueTag is not inlined by the compiler. Copying this
 		// special case into one of the EncodeFooValue functions speeds it up by
 		// ~4ns.
 		return append(appendTo, byte(typ))
 	}
-	return EncodeNonsortingUvarint(appendTo, uint64(colID)<<4|uint64(typ))
+	return EncodeNonsortingUvarint(appendTo, uint64(colIDDelta)<<4|uint64(typ))
 }
 
 // EncodeNullValue encodes a null value, appends it to the supplied buffer, and
 // returns the final buffer.
-func EncodeNullValue(appendTo []byte, colID uint32) []byte {
-	return EncodeValueTag(appendTo, colID, Null)
+func EncodeNullValue(appendTo []byte, colIDDelta uint32) []byte {
+	return EncodeValueTag(appendTo, colIDDelta, Null)
 }
 
 // EncodeNotNullValue encodes a not null value, appends it to the supplied
 // buffer, and returns the final buffer.
-func EncodeNotNullValue(appendTo []byte, colID uint32) []byte {
-	return EncodeValueTag(appendTo, colID, NotNull)
+func EncodeNotNullValue(appendTo []byte, colIDDelta uint32) []byte {
+	return EncodeValueTag(appendTo, colIDDelta, NotNull)
 }
 
 // EncodeBoolValue encodes a bool value, appends it to the supplied buffer, and
 // returns the final buffer.
-func EncodeBoolValue(appendTo []byte, colID uint32, b bool) []byte {
+func EncodeBoolValue(appendTo []byte, colIDDelta uint32, b bool) []byte {
 	if b {
-		return EncodeValueTag(appendTo, colID, True)
+		return EncodeValueTag(appendTo, colIDDelta, True)
 	}
-	return EncodeValueTag(appendTo, colID, False)
+	return EncodeValueTag(appendTo, colIDDelta, False)
 }
 
 // EncodeIntValue encodes an int value with its value tag, appends it to the
 // supplied buffer, and returns the final buffer.
-func EncodeIntValue(appendTo []byte, colID uint32, i int64) []byte {
-	appendTo = EncodeValueTag(appendTo, colID, Int)
+func EncodeIntValue(appendTo []byte, colIDDelta uint32, i int64) []byte {
+	appendTo = EncodeValueTag(appendTo, colIDDelta, Int)
 	return EncodeUntaggedIntValue(appendTo, i)
 }
 
@@ -2617,8 +2617,8 @@ const floatValueEncodedLength = uint64AscendingEncodedLength
 
 // EncodeFloatValue encodes a float value with its value tag, appends it to the
 // supplied buffer, and returns the final buffer.
-func EncodeFloatValue(appendTo []byte, colID uint32, f float64) []byte {
-	appendTo = EncodeValueTag(appendTo, colID, Float)
+func EncodeFloatValue(appendTo []byte, colIDDelta uint32, f float64) []byte {
+	appendTo = EncodeValueTag(appendTo, colIDDelta, Float)
 	return EncodeUntaggedFloatValue(appendTo, f)
 }
 
@@ -2636,8 +2636,8 @@ func EncodeUntaggedFloat32Value(appendTo []byte, f float32) []byte {
 
 // EncodeBytesValue encodes a byte array value with its value tag, appends it to
 // the supplied buffer, and returns the final buffer.
-func EncodeBytesValue(appendTo []byte, colID uint32, data []byte) []byte {
-	appendTo = EncodeValueTag(appendTo, colID, Bytes)
+func EncodeBytesValue(appendTo []byte, colIDDelta uint32, data []byte) []byte {
+	appendTo = EncodeValueTag(appendTo, colIDDelta, Bytes)
 	return EncodeUntaggedBytesValue(appendTo, data)
 }
 
@@ -2650,15 +2650,15 @@ func EncodeUntaggedBytesValue(appendTo []byte, data []byte) []byte {
 
 // EncodeArrayValue encodes a byte array value with its value tag, appends it to
 // the supplied buffer, and returns the final buffer.
-func EncodeArrayValue(appendTo []byte, colID uint32, data []byte) []byte {
-	appendTo = EncodeValueTag(appendTo, colID, Array)
+func EncodeArrayValue(appendTo []byte, colIDDelta uint32, data []byte) []byte {
+	appendTo = EncodeValueTag(appendTo, colIDDelta, Array)
 	return EncodeUntaggedBytesValue(appendTo, data)
 }
 
 // EncodeTimeValue encodes a time.Time value with its value tag, appends it to
 // the supplied buffer, and returns the final buffer.
-func EncodeTimeValue(appendTo []byte, colID uint32, t time.Time) []byte {
-	appendTo = EncodeValueTag(appendTo, colID, Time)
+func EncodeTimeValue(appendTo []byte, colIDDelta uint32, t time.Time) []byte {
+	appendTo = EncodeValueTag(appendTo, colIDDelta, Time)
 	return EncodeUntaggedTimeValue(appendTo, t)
 }
 
@@ -2671,8 +2671,8 @@ func EncodeUntaggedTimeValue(appendTo []byte, t time.Time) []byte {
 
 // EncodeTimeTZValue encodes a timetz.TimeTZ value with its value tag, appends it to
 // the supplied buffer, and returns the final buffer.
-func EncodeTimeTZValue(appendTo []byte, colID uint32, t timetz.TimeTZ) []byte {
-	appendTo = EncodeValueTag(appendTo, colID, TimeTZ)
+func EncodeTimeTZValue(appendTo []byte, colIDDelta uint32, t timetz.TimeTZ) []byte {
+	appendTo = EncodeValueTag(appendTo, colIDDelta, TimeTZ)
 	return EncodeUntaggedTimeTZValue(appendTo, t)
 }
 
@@ -2685,14 +2685,14 @@ func EncodeUntaggedTimeTZValue(appendTo []byte, t timetz.TimeTZ) []byte {
 
 // EncodeVoidValue encodes a void with its value tag, appends it to
 // the supplied buffer and returns the final buffer.
-func EncodeVoidValue(appendTo []byte, colID uint32) []byte {
-	return EncodeValueTag(appendTo, colID, Void)
+func EncodeVoidValue(appendTo []byte, colIDDelta uint32) []byte {
+	return EncodeValueTag(appendTo, colIDDelta, Void)
 }
 
 // EncodeBox2DValue encodes a geopb.BoundingBox with its value tag, appends it to
 // the supplied buffer and returns the final buffer.
-func EncodeBox2DValue(appendTo []byte, colID uint32, b geopb.BoundingBox) ([]byte, error) {
-	appendTo = EncodeValueTag(appendTo, colID, Box2D)
+func EncodeBox2DValue(appendTo []byte, colIDDelta uint32, b geopb.BoundingBox) ([]byte, error) {
+	appendTo = EncodeValueTag(appendTo, colIDDelta, Box2D)
 	return EncodeUntaggedBox2DValue(appendTo, b)
 }
 
@@ -2708,8 +2708,8 @@ func EncodeUntaggedBox2DValue(appendTo []byte, b geopb.BoundingBox) ([]byte, err
 
 // EncodeGeoValue encodes a geopb.SpatialObject value with its value tag, appends it to
 // the supplied buffer, and returns the final buffer.
-func EncodeGeoValue(appendTo []byte, colID uint32, so *geopb.SpatialObject) ([]byte, error) {
-	appendTo = EncodeValueTag(appendTo, colID, Geo)
+func EncodeGeoValue(appendTo []byte, colIDDelta uint32, so *geopb.SpatialObject) ([]byte, error) {
+	appendTo = EncodeValueTag(appendTo, colIDDelta, Geo)
 	return EncodeUntaggedGeoValue(appendTo, so)
 }
 
@@ -2725,8 +2725,8 @@ func EncodeUntaggedGeoValue(appendTo []byte, so *geopb.SpatialObject) ([]byte, e
 
 // EncodeDecimalValue encodes an apd.Decimal value with its value tag, appends
 // it to the supplied buffer, and returns the final buffer.
-func EncodeDecimalValue(appendTo []byte, colID uint32, d *apd.Decimal) []byte {
-	appendTo = EncodeValueTag(appendTo, colID, Decimal)
+func EncodeDecimalValue(appendTo []byte, colIDDelta uint32, d *apd.Decimal) []byte {
+	appendTo = EncodeValueTag(appendTo, colIDDelta, Decimal)
 	return EncodeUntaggedDecimalValue(appendTo, d)
 }
 
@@ -2749,8 +2749,8 @@ func EncodeUntaggedDecimalValue(appendTo []byte, d *apd.Decimal) []byte {
 
 // EncodeDurationValue encodes a duration.Duration value with its value tag,
 // appends it to the supplied buffer, and returns the final buffer.
-func EncodeDurationValue(appendTo []byte, colID uint32, d duration.Duration) []byte {
-	appendTo = EncodeValueTag(appendTo, colID, Duration)
+func EncodeDurationValue(appendTo []byte, colIDDelta uint32, d duration.Duration) []byte {
+	appendTo = EncodeValueTag(appendTo, colIDDelta, Duration)
 	return EncodeUntaggedDurationValue(appendTo, d)
 }
 
@@ -2764,8 +2764,8 @@ func EncodeUntaggedDurationValue(appendTo []byte, d duration.Duration) []byte {
 
 // EncodeBitArrayValue encodes a bit array value with its value tag,
 // appends it to the supplied buffer, and returns the final buffer.
-func EncodeBitArrayValue(appendTo []byte, colID uint32, d bitarray.BitArray) []byte {
-	appendTo = EncodeValueTag(appendTo, colID, BitArray)
+func EncodeBitArrayValue(appendTo []byte, colIDDelta uint32, d bitarray.BitArray) []byte {
+	appendTo = EncodeValueTag(appendTo, colIDDelta, BitArray)
 	return EncodeUntaggedBitArrayValue(appendTo, d)
 }
 
@@ -2784,8 +2784,8 @@ func EncodeUntaggedBitArrayValue(appendTo []byte, d bitarray.BitArray) []byte {
 
 // EncodeUUIDValue encodes a uuid.UUID value with its value tag, appends it to
 // the supplied buffer, and returns the final buffer.
-func EncodeUUIDValue(appendTo []byte, colID uint32, u uuid.UUID) []byte {
-	appendTo = EncodeValueTag(appendTo, colID, UUID)
+func EncodeUUIDValue(appendTo []byte, colIDDelta uint32, u uuid.UUID) []byte {
+	appendTo = EncodeValueTag(appendTo, colIDDelta, UUID)
 	return EncodeUntaggedUUIDValue(appendTo, u)
 }
 
@@ -2797,8 +2797,8 @@ func EncodeUntaggedUUIDValue(appendTo []byte, u uuid.UUID) []byte {
 
 // EncodeIPAddrValue encodes a ipaddr.IPAddr value with its value tag, appends
 // it to the supplied buffer, and returns the final buffer.
-func EncodeIPAddrValue(appendTo []byte, colID uint32, u ipaddr.IPAddr) []byte {
-	appendTo = EncodeValueTag(appendTo, colID, IPAddr)
+func EncodeIPAddrValue(appendTo []byte, colIDDelta uint32, u ipaddr.IPAddr) []byte {
+	appendTo = EncodeValueTag(appendTo, colIDDelta, IPAddr)
 	return EncodeUntaggedIPAddrValue(appendTo, u)
 }
 
@@ -2811,32 +2811,32 @@ func EncodeUntaggedIPAddrValue(appendTo []byte, u ipaddr.IPAddr) []byte {
 // EncodeJSONValue encodes an already-byte-encoded JSON value with no value tag
 // but with a length prefix, appends it to the supplied buffer, and returns the
 // final buffer.
-func EncodeJSONValue(appendTo []byte, colID uint32, data []byte) []byte {
-	appendTo = EncodeValueTag(appendTo, colID, JSON)
+func EncodeJSONValue(appendTo []byte, colIDDelta uint32, data []byte) []byte {
+	appendTo = EncodeValueTag(appendTo, colIDDelta, JSON)
 	return EncodeUntaggedBytesValue(appendTo, data)
 }
 
 // EncodeTSQueryValue encodes an already-byte-encoded TSQuery value with no
 // value tag but with a length prefix, appends it to the supplied buffer, and
 // returns the final buffer.
-func EncodeTSQueryValue(appendTo []byte, colID uint32, data []byte) []byte {
-	appendTo = EncodeValueTag(appendTo, colID, TSQuery)
+func EncodeTSQueryValue(appendTo []byte, colIDDelta uint32, data []byte) []byte {
+	appendTo = EncodeValueTag(appendTo, colIDDelta, TSQuery)
 	return EncodeUntaggedBytesValue(appendTo, data)
 }
 
 // EncodeTSVectorValue encodes an already-byte-encoded TSVector value with no
 // value tag but with a length prefix, appends it to the supplied buffer, and
 // returns the final buffer.
-func EncodeTSVectorValue(appendTo []byte, colID uint32, data []byte) []byte {
-	appendTo = EncodeValueTag(appendTo, colID, TSVector)
+func EncodeTSVectorValue(appendTo []byte, colIDDelta uint32, data []byte) []byte {
+	appendTo = EncodeValueTag(appendTo, colIDDelta, TSVector)
 	return EncodeUntaggedBytesValue(appendTo, data)
 }
 
 // EncodePGVectorValue encodes an already-byte-encoded PGVector value with no
 // value tag but with a length prefix, appends it to the supplied buffer, and
 // returns the final buffer.
-func EncodePGVectorValue(appendTo []byte, colID uint32, data []byte) []byte {
-	appendTo = EncodeValueTag(appendTo, colID, PGVector)
+func EncodePGVectorValue(appendTo []byte, colIDDelta uint32, data []byte) []byte {
+	appendTo = EncodeValueTag(appendTo, colIDDelta, PGVector)
 	return EncodeUntaggedBytesValue(appendTo, data)
 }
 
@@ -2848,12 +2848,12 @@ func EncodePGVectorValue(appendTo []byte, colID uint32, data []byte) []byte {
 // PeekValueLength and each of the DecodeFooValue methods will still work as
 // expected with `b[typeOffset:]`. (Except, obviously, the column id is no
 // longer encoded so if this suffix is passed back to DecodeValueTag, the
-// returned colID should be discarded.)
+// returned colIDDelta should be discarded.)
 //
 // Concretely:
 //
 //	b := ...
-//	typeOffset, _, colID, typ, err := DecodeValueTag(b)
+//	typeOffset, _, colIDDelta, typ, err := DecodeValueTag(b)
 //	_, _, _, typ, err := DecodeValueTag(b[typeOffset:])
 //
 // will return the same typ and err and
@@ -2863,7 +2863,9 @@ func EncodePGVectorValue(appendTo []byte, colID uint32, data []byte) []byte {
 //
 // will return the same thing. PeekValueLength works as expected with either of
 // `b` or `b[typeOffset:]`.
-func DecodeValueTag(b []byte) (typeOffset int, dataOffset int, colID uint32, typ Type, err error) {
+func DecodeValueTag(
+	b []byte,
+) (typeOffset int, dataOffset int, colIDDelta uint32, typ Type, err error) {
 	// TODO(dan): This can be made faster by special casing the single byte
 	// version and skipping the column id extraction when it's not needed.
 	if len(b) == 0 {
@@ -2875,7 +2877,7 @@ func DecodeValueTag(b []byte) (typeOffset int, dataOffset int, colID uint32, typ
 	if err != nil {
 		return 0, 0, 0, Unknown, err
 	}
-	colID = uint32(tag >> 4)
+	colIDDelta = uint32(tag >> 4)
 
 	typ = Type(tag & 0xf)
 	typeOffset = n - 1
@@ -2888,7 +2890,7 @@ func DecodeValueTag(b []byte) (typeOffset int, dataOffset int, colID uint32, typ
 		typ = Type(tag)
 		dataOffset += n
 	}
-	return typeOffset, dataOffset, colID, typ, nil
+	return typeOffset, dataOffset, colIDDelta, typ, nil
 }
 
 // DecodeBoolValue decodes a value encoded by EncodeBoolValue.

